rust/worker/src/execution/operators/merge_metadata_results.rs

```diff
@@ -107,289 +113,136 @@ impl Operator<MergeMetadataResultsOperatorInput, MergeMetadataResultsOperatorOut
-                                        *e,
-                                    ),
-                                );
-                            }
-                            RecordSegmentReaderCreationError::InvalidNumberOfFiles => {
-                                error!("Error creating Record Segment: {:?}", e);
-                                return Err(
-                                    MergeMetadataResultsOperatorError::RecordSegmentCreationError(
-                                        *e,
-                                    ),
-                                );
-                            }
-                        }
-                    }
-                };
-                for merged_id in merged_ids {
-                    // Skip already taken from log.
-                    if visited_ids.contains(&merged_id) {
-                        continue;
-                    }
-
-                    let user_id = match record_segment_reader
-                        .get_user_id_for_offset_id(merged_id)
-                        .await
-                    {
-                        Ok(user_id) => user_id,
-                        Err(e) => {
-                            println!("Error reading Record Segment: {:?}", e);
-                            return Err(MergeMetadataResultsOperatorError::RecordSegmentReadError);
-                        }
-                    };
-                    ids.push(user_id.to_string());
+        let mut logged_offset_ids: HashSet<u32> = HashSet::new();
+        if input.include_metadata {
+            metadata = vec![None; take_count];
+            documents = vec![None; take_count];
+        }
 
-                    if input.include_metadata {
-                        let record = match record_segment_reader
-                            .get_data_for_offset_id(merged_id)
-                            .await
-                        {
-                            Ok(record) => record,
-                            Err(e) => {
-                                tracing::error!("Error reading Record Segment: {:?}", e);
-                                return Err(
-                                    MergeMetadataResultsOperatorError::RecordSegmentReadError,
-                                );
-                            }
-                        };
-                        metadata.push(record.metadata.clone());
-                        match record.document {
-                            Some(document) => documents.push(Some(document.to_string())),
-                            None => documents.push(None),
-                        }
-                    }
+        // Hydrate the data from the materialized logs first
+        for (log, _) in mat_records.iter() {
+            // It's important to account for the records that are
+            // deleted also here so that we can subsequently ignore
+            // them when reading the record segment.
+            logged_offset_ids.insert(log.offset_id);
```

Comment:
`logged_offset_ids` does not have to contain all the log entries; you could make it contain only entries that are both in the log and in the `truncated_offset_id_order`. Also remember to include the deleted entries in this case. Again see the original code for more details

Reply:
Yes, that could be an optimization.